### ISSUE TITLE
Improve the REPL

### ISSUE THREAD

**Issue by ice1000** [aya-prover/aya-dev#87]
Here's a summary of what's missing from the current impl:

+ [ ] Lots of methods and fields are package-private, and I don't know why. I guess making more of them `private` is a better idea.
+ [ ] Parser is unimplemented.
+ [ ] Completion is unimplemented. I added completion for keywords, but that's only available for non-commands. If your line starts with `:`, then no completion. This should be changed after we have commands that accept expressions.
+ [ ] No repl commands. I think we need a general interface for repl commands so the repl framework is extensible. Here's a list of what I'm looking for.
  + [ ] `:t`, `:type` from GHCi. Completion should be enabled for this command.
  + [ ] `:normalize` or `:eval` -- whatever you'd like to call it -- for normalizing the expression. Completion should be enabled.
  + [ ] `:prompt` that changes the prompt.
  + [ ] An option to change the pretty printing width.
  + [ ] An option to change pretty printing options. Available options should be completed.
+ [ ] Save history for jline3 based repl. We may start using `~/.aya/something` for this purpose.
+ [ ] Handle redefinitions.
+ [ ] Completion of commands. This is very important.

_Originally posted by @ice1000 in https://github.com/aya-prover/aya-dev/issues/86#issuecomment-944882637_

**Comment by ice1000** [comment]
@refparo How would you expect, say, to change a code using `~/.aya/something` to a code using XGD base dir? I looked up the link and I expected 'read $XDG_STATE_HOME, and if it's null, use `~/.aya`, am I right?

**Comment by refparo** [comment]
> If `$XDG_STATE_HOME` is either not set or empty, a default equal to `$HOME/.local/state` should be used.

According to [the specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html).

**Comment by ice1000** [comment]
@refparo Are you sure that everyone wants to use `$HOME/.local/state/something`? What if two different programs both use, say, `$HOME/.local/state/history`?

**Comment by ice1000** [comment]
Or should I create a subdir under `$HOME/.local/state`, like `$HOME/.local/state/aya`?

**Comment by refparo** [comment]
Yeah, use `$HOME/.local/state/aya_history`, or `$HOME/.local/state/aya/history`. It's just like how people use `$HOME/.config` (which is the `$XDG_CONFIG_HOME`)

**Comment by ice1000** [comment]
Great

**Comment by ice1000** [comment]
@ShreckYe It won't work. Your command framework should be able to set the prompt of a repl, do you think picocli knows about our repl?

**Comment by ice1000** [comment]
Also, cli completion is provided by jline3. I think doing completion in the same application based on different frameworks will lead to some inconsistency.

But you have the opportunity to try things out, let me know if it works

**Comment by ice1000** [comment]
> code completion is done in Arend with ANTLR right?

You mean arend repl? No, I wrote it from scratch. So it's quite elementary.

**Comment by ShreckYe** [comment]
OK. I think it is doable with ANTLR, isn't it?

**Comment by ice1000** [comment]
> OK. I think it is doable with ANTLR, isn't it?

Sure, but can you do the command framework first? I want to implement some interesting commands

**Comment by ShreckYe** [comment]
OK.

**Comment by ice1000** [comment]
@ShreckYe Also, I am thinking about integrating the repl in the lsp. Do you think it's possible? Like, provide some vscode-specific commands when the repl is invoked in the lsp

**Comment by ShreckYe** [comment]
OK I will give it a try.

**Comment by imkiva** [comment]
For where to store aya config, history or cache files. I suggest one of them:
1. Use our own base path, in every platform, like OpenSSH (`~/.ssh`)
2. Use our own base path, in every platform, but provide options to switch to other dirs, like cargo, rustup, and gradle via env `XXX_HOME`. We can support XDG Base Directory for **desktop users** by `export AYA_HOME="$XDG_DATA_HOME/aya"` or a command line option.

I don't suggest:
1. Use `~/Library/Application Support` on macOS. This path is used for applications that ship with an app bundle `.app` file. We may encounter some strange permission problems. I didn't see any command line only programs use this path. And the path contains an unfriendly space that adds burden to command-line operation.
2. Use all XDG base directories. This means we save different files in different dirs, which is tiring for both users and developers. And we also need to manually split workspace content into configs, downloads, etc.

Why is our own path that important?
For users that use all main systems (like me), **it is important and ideal to easily get re-started from experiences before**. I can know that in every platform If I want to find my ssh keys, I just need to navigate to `~/.ssh` instead of googling something like `ssh key location on windows`. 

Why shouldn't use XDG base directories by default?
- Aya can be run on server-side. Using desktop-specific dir looks strange.
- Some servers didn't even have `XDG_XXX_HOME` set.

Then how to use XDG base directories?
By environmental variables or command-line options, as said above. Actually, lots of programs support XDGs the same way, according to https://wiki.archlinux.org/title/XDG_Base_Directory

**Comment by ice1000** [comment]
@imkiva Great!

**Comment by imkiva** [comment]
Two suggestions:

- Use `<stdin>` as file name, instead of `Option.none()` which leads to `<unknown-file>`
- Support exiting REPL by Ctrl-D which sends EOF to stdin, currently we will show a stack trace which is unexpected

**Comment by imkiva** [comment]
me: "Which one is important to me? \<stdin\>? \<interactive\>?"
\<unknown-file\>: "\<stdin\>"
me: "Think again"
\<unknown-file\>: "Then \<interactive\>"
me: "Think again"
\<unknown-file\>: "Still \<stdin\>"
me: "None of them matters. It matters to me without \<unknown-file\>."
